### ISSUE TITLE
Built-in UI: No mouse pointer on macOS on startup

### ISSUE THREAD

**Issue by ChoccyHobNob** [mamedev/mame#1232]
When MAME opens on macOS I have no mouse pointer. I have to enter a sub menu, pick an option, come back out and then go back to the main menu to force a reinitialization for the mouse to be available, then it works fine.

**Comment by cuavas** [comment]
I think it has to do with two unintialised members in the `sdl_window_info` class.  I've raised this with @bradhugh who last refactored the file and is responsible for the uninitialised members, but he seems to be MIA again.  I'll try initialising them to false and see if it fixes it.

**Comment by cuavas** [comment]
Can you see if a2b99f814965f8e703b380a6c67b4f66f1a18749 fixes it?  Preferably just apply that over what you've got to isolate the change.

**Comment by ChoccyHobNob** [comment]
It didn't fix the problem but is has had a positive effect in some of the erratic behavior from the mouse. There are a whole bunch of mouse issues with the SDL build, I've noticed on both OS X and Raspberry Pi. 

Sometimes when starting MAME there will be no mouse pointer at all (80%~), sometimes the orange pointer will appear but wont move at all (15%~), sometimes you will get the system mouse pointer over the orange pointer and it will be frozen (4%~), sometimes it will start up with the system mouse pointer over the orange pointer and it will move ok (1%~). The times the orange pointer appears but is frozen, the system mouse pointer is on screen at the same coords before MAME initializes, sometimes it disappears, sometimes not. When returning from a menu or a game the orange pointer is always in the bottom right corner and then works ok.

When you select a game that uses the mouse as a controller, sometimes you will get the system mouse on top of the display moving in sync with the game cursor (10%~), sometimes it wont appear at all (90%~). On the Raspberry Pi you can't use the menu it takes forever to appear and is too slow to use but when running a game directly from the command line sometimes the system mouse pointer will appear in the top left corner, and will move with the mouse when running a game, even though there is no window manager running.

Change a2b99f8 seems to have fixed the problem on OS X with the system pointer appearing over the orange pointer so it is now at least consistent. With the change now, 100% of the time when starting there is no orange pointer and no system pointer. After running a game and returning, or entering a configuration menu and returning, 100% of the time now the orange pointer appears and there is no system pointer. 100% of the time in a game there is no system pointer in sync with the game cursor. I haven't tried this on the Raspberry Pi.

So... this fixed a whole lot of odd mouse behavior but didn't fix the problem I reported!

**Comment by Headrush69** [comment]
This patch greatly improved my mouse issues also, but problems remain.

Fixed: System mouse appearing on screen and trackball movements restricted by screen bounds.

Issue: For mouse to work, I either have to start NEWUI and select game, or when directly starting a game from command line, than go into the in game menu and than mouse works.

**Comment by bradhugh** [comment]
@cuavas just checked in a change to properly initialize members as he mentioned in commit a2b99f81.  can you check if this helped?

**Comment by bradhugh** [comment]
Ok, nevermind - sorry I'm just catching up.  I see this did fix most of your strange inconsistencies, but you now don't have the any pointer at all when staring up.  I'll see if I can repro this and if so, track it down for you unless @cuavas is already working on it.

**Comment by bradhugh** [comment]
I couldn't repro this at all.  It's working consistently for me.  Are you running full screen or windowed?  What OS and compiler version for OSX?

**Comment by ChoccyHobNob** [comment]
Full screen. If I use a window it works consistently.
OS X 10.11.6
Clang 703.0.31 (XCode 7.3.1)

I just did this...
`cd ~/my/mame/path
./mame -verbose >verbose.txt`

- screen opens with "initializing" in the middle
- ui screen opens, no pointer 
- select pacman
- info screen comes up, hit esc
- screen opens with "initializing" in the middle
- ui screen opens, pointer is off the sceen at 0,0 (bottom right hand corner) and moves correctly
- hit esc to quit

Here's the verbose.txt it generated
[verbose.txt](https://github.com/mamedev/mame/files/427521/verbose.txt)

Happens like this 100% of the time now, change a2b99f8 seems to have fixed all the weird inconsistencies.

**Comment by bradhugh** [comment]
Thanks for the information. I think this has to have something to do with the code in the UI layer that determines whether or not it should show the UI mouse cursor.  There is code that checks whether the handler is a menu to determine whether to show the mouse.

Can you clone, build, and run a test from my repro?  I added some tracing in the branch "ui_mouse_debug" below:
https://github.com/bradhugh/mame/tree/ui_mouse_debug

If you can run this with -verbose, I think it should show what is actually going on. Let me know if you can do this.

**Comment by ChoccyHobNob** [comment]
Hi Brad, yes of course I can do that, I'm grabbing it now, I'll let you know when I've run it. Thanks for all your effort on this btw!

**Comment by ChoccyHobNob** [comment]
OK here's the new verbose output

[verbose2.txt](https://github.com/mamedev/mame/files/432564/verbose2.txt)

**Comment by bradhugh** [comment]
I looked at this and can't seem to see any difference in the tracing between yours and mine with the added lines.  I did however see one line in your verbose output that's not in mine and I can't explain.  In your output there is this sequence:
Starting No Driver Loaded ':'
Changing callback type: GENERAL -> MENU
Changing callback type: MENU -> GENERAL
Changing callback type: GENERAL -> MENU
**Starting No Driver Loaded**
Detected is_menu_active change to: active
Detected UI Mouse render choice change to: true
The second "starting" message there is coming from something that I'm clearly not running and I could not find it in source.  Do you have any cheats or LUA add-ins enabled?

**Comment by ChoccyHobNob** [comment]
I have hiscore enabled in plugin.ini, I just tried disabling it and it made no difference to the mouse behavior. Running -verbose for both states does cause an extra "Starting No Driver Loaded" to be output. 

What version of SDL2 are you using?

**Comment by ChoccyHobNob** [comment]
On a hunch I just tried running `./mame arkanoid` and it runs but only responds to cursor keys, the mouse doesn't work. 

If I run the ui and select arkanoid, it runs and the mouse controls the vaus as expected. 

I wonder if the mouse pointer is being drawn correctly at position 0,0 when the menu is first run but it isn't being polled at all so it never moves. Then it's not until MAME is reinitialized by either running a game or changing the settings and returning to the menu that the mouse is being read and responds to movement.

**Comment by bradhugh** [comment]
That was it.  The version of SDL!  I can repro now with the "public" version from their web site.  I had downloaded and built a recent version from their mercurial repository a month or two ago and I bet it has fixes for this.  Let me see what I can figure out now that I repro.

**Comment by bradhugh** [comment]
Ok, I've pushed a test fix for you to this same repo.  Can you pull the latest from my repo ui_mouse_debug branch, build, and see if it's fixed?

**Comment by ChoccyHobNob** [comment]
That's now almost right! When started without a game, the mouse appears in the top left. This is consistent and works every time. Running a mouse controlled game, everything works as expected!

Small bug still left when running without the menu (`./mame opwolf`). Where-ever the system pointer is at the time you run the command becomes the top left. you cant move the mouse up'er or left'er than that point. Try putting the mouse pointer in the middle of the screen and then typing `./mame opwolf` you will only be able to move the crosshairs in the bottom quarter of the screen.

**Comment by bradhugh** [comment]
Ok, great catch. I'll take a look and see what I can figure out.

**Comment by bradhugh** [comment]
I'm having trouble finding a workaround. I just tested again with my newer build of SDL against the released version of MAME and not only does the UI work well, but the operation wolf issue isn't able to be reproduced either.  What are your thoughts on just recommending Mac users build a newer version of SDL if they have this problem?  I'll continue to look for a workaround, but no luck yet.

**Comment by rb6502** [comment]
If that's necessary, I'd like to build something as close as possible to the binary package distributed by SDL.org where you can drag it into /Library/Frameworks.  Requiring Mac users to actually build SDL is something I really want to avoid, even if there's a nice Xcode project for it.

**Comment by bradhugh** [comment]
[SDL2.framework.zip](https://github.com/mamedev/mame/files/437930/SDL2.framework.zip)
Sure, I just built this which is from the latest unmodified source from SDL mercurial. Can one of you guys validate that it resolves it?

**Comment by ChoccyHobNob** [comment]
I've also just built an SDL2.framework (and a standard .dmg from that, which I've attached so we can all be on the same code) but I'm still having issues. I'm just doing some testing so I'll come back in a while but it looks like the new SDL2 won't fix the problem.

Brad, the mouse being stuck in the bottom quarter of the screen is a bit of a red herring. It's stuck there regardless of where the mouse pointer is when it's run, but the area does expand to where the mouse pointer is. Wow this is complicated to describe! Let me do some more testing with both versions of SDL2 and both versions of MAME and I'll report back.

[SDL2.dmg.zip](https://github.com/mamedev/mame/files/437962/SDL2.dmg.zip)

**Comment by bradhugh** [comment]
I know what you mean regarding the mouse in the quarter.  It does expand, but never fully.  I was never able to get it to go the full vertical height of the screen even once the movable area expanded.

**Comment by bradhugh** [comment]
I just tried your DMG with the latest MAME git sources and I can't repro the problem.  It's fixed - same as with mine.  If you can still repro, make sure you're using the mame build off of master, not the one where I attempted to fix it.

**Comment by ChoccyHobNob** [comment]
With or without @cuavas change a2b99f8?

**Comment by ChoccyHobNob** [comment]
OK I've done 3 versions of mame with 2 versions of SDL2 with 3 runs for each

`./mame`
`./mame` -> pick opwolf -> exit
`./mame opwolf

You're right, SDL2 fixes it as long as I'm not using the version where you tried to fix it. 

Interesting to note is that @cuavas fix appears to still be an improvement. Without it the system pointer is on top of the ui pointer and sometimes it's there over the game crosshair as well. With it all that inconsistency goes away.

So I'm saying that change a2b99f8 and a new version of SDL2 fixes the problem.

**Comment by bradhugh** [comment]
Yes, that's my experience as well.  For now, I would like to leave the code as-is since it appears to be an SDL bug and updating SDL fixes it.  If I can find a perfect workaround, I'll implement it, but for now leaving it as-is and recommend mac users use an updated SDL.

**Comment by ChoccyHobNob** [comment]
There are similar issues with linux builds too. I shall try a new version of SDL2 on there too.

**Comment by bradhugh** [comment]
I tested the same scenarios on Ubuntu 16.04 LTS with whatever version of SDL you normally get through apt-get and didn't have any issues.

**Comment by ChoccyHobNob** [comment]
I think that's 2.0.2, it is for Debian Jessie. It's probably just the Raspberry Pi people who will be using 2.0.4 (it includes hardware support for the Pi's GPU), Retropie definitely rolls out a version.

**Comment by ChoccyHobNob** [comment]
I've uploaded the SDL-2.0.4-10216.dmg to [here](http://choccyhobnob.com/software/mame/) ready for 0.177's release. Someone should let r0ni know so that he can do the same.

**Comment by jloco** [comment]
Is this 2.0.4 build newer than the released one on sdl.org?  (Just checking in to verify things before updating mame next week) I will post the package if it is, so users can install it at the same time as the next release. And post some kind of notice on the site as well to alert users. 

Jay (r0ni)

**Comment by ChoccyHobNob** [comment]
Yes this one was built yesterday from the mercurial. The one on sdl.org dates back to January

**Comment by bradhugh** [comment]
So I've got a code workaround that is working for me. Can you try out the latest from my repo's ui_mouse_debug branch?  I found a few things. 1) the problem *only* manifests on 2.0.4. It's not there in 2.0.2 or 2.0.3.  2) If I pass the SDL hint to use relative mouse in warp mode, it fixes it.
So the code fix is that if we detect the buggy SDL version, we use warp mode instead of the raw-event based relative mode.  Please let me know if it works for you.

**Comment by ChoccyHobNob** [comment]
Yep that works for me with the current stable version of SDL2-2.0.4 and is still fine with the new version we created yesterday. 

I still want to try this on a Raspberry Pi. My dev one is busy right now but as soon as I can I'll try it on there too and see if it fixes the odd problems with mice that I have there.

**Comment by bradhugh** [comment]
This member init probably helps significantly, but this latest part will have no effect as its scoped to mac.
